src/components/data_comps/dshr_nuopc/dshr_dfield_mod.F90

```diff
@@ -0,0 +1,292 @@
+    dfield_new%sdat_avect_index = iunset
+    dfield_new%stream_data1d => null()
+
+    ! loop over all input streams and ! determine if the strm_fld is in the attribute vector of stream ns
+    do ns = 1, sdat%nstreams
+       ! determine if the strm_fld is in the attribute vector of stream ns
+       kf  = mct_aVect_indexRA(sdat%avs(ns), trim(strm_fld), perrWith='quiet')
+       if (kf > 0) then
+          dfield_new%sdat_stream_index = ns
+          dfield_new%sdat_avect_index = kf
+          allocate(dfield_new%stream_data1d(lsize))
+          strm_ptr => dfield_new%stream_data1d
+          if (present(logunit) .and. present(masterproc)) then
+             if (masterproc) then
+                write(logunit,*)'(dshr_addfield_add) allocating memory for stream field strm_'//trim(strm_fld)
+             end if
+          end if
+          exit
+       end if
+    end do
+
+  end subroutine dshr_dfield_add_strmfld
+
+!===============================================================================
+
+  subroutine dshr_dfield_add_1d(dfields, sdat, state_fld, strm_fld, state, &
+       state_ptr, strm_ptr, logunit, masterproc, rc)
+
+    ! Set 1d dfield values
+
+    type(dfield_type)      , pointer       :: dfields
+    type(shr_strdata_type) , intent(in)    :: sdat
+    character(len=*)       , intent(in)    :: state_fld
+    character(len=*)       , intent(in)    :: strm_fld
+    type(ESMF_State)       , intent(inout) :: state
+    real(r8), optional     , pointer       :: state_ptr(:)
+    real(r8), optional     , pointer       :: strm_ptr(:)
+    integer , optional     , intent(in)    :: logunit
+    logical , optional     , intent(in)    :: masterproc
+    integer                , intent(out)   :: rc
+
+    ! local variables
+    type(dfield_type), pointer :: dfield_new
+    integer :: ns, kf, lsize
+    character(len=*), parameter :: subname='(dfield_add_1d)'
+    ! ----------------------------------------------
+
+    rc = ESMF_SUCCESS
+
+    allocate(dfield_new)
+    dfield_new%next => dfields
+    dfields => dfield_new
+
+    ! determine local size
+    lsize = mct_avect_lsize(sdat%avs(1))
+
+    ! Initialize strm_ptr and state_ptr if it is present
+    ! These will be set to valid values if the relevant fields are found
+    if (present(strm_ptr)) strm_ptr => null()
+    if (present(state_ptr)) state_ptr => null()
+    dfield_new%sdat_stream_index = iunset
+    dfield_new%sdat_avect_index = iunset
+
+    ! always allocate memory for stream_data and initialize it to 0
+    ! note that if the attribute vector is not found in the streams
+
+    ! loop over all input streams
+    do ns = 1, sdat%nstreams
+       ! determine if the strm_fld is in the attribute vector of stream ns
+       kf  = mct_aVect_indexRA(sdat%avs(ns), trim(strm_fld), perrWith='quiet')
+       if (kf > 0) then
+          dfield_new%sdat_stream_index = ns
+          dfield_new%sdat_avect_index = kf
+          allocate(dfield_new%stream_data1d(lsize))
```

Comment:
Allocate statements should use the optional stat variable and explicitly check for errors.